[Darren159] iP

Darren159 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/Darren159/ip/releases/tag/v0.1).
2. run java -jar duke.jar.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is __FREE__!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Darren159 committed:
Rename chatbot and add greeting and exit messages

Darren159 committed:
Add command echoing and exiting

Darren159 committed:
Add ability to store text entered by user and display them back to the user

Darren159 committed:
Add ability to mark tasks as done

Darren159 committed:
Add support for tracking three types of tasks

Darren159 committed:
Add Automated Text UI Testing

Darren159 committed:
Add error handling

Darren159 committed:
Add support for deleting tasks from the list

Darren159 committed:
Use Enumerations

Darren159 committed:
Add support for saving and loading tasks from hard disk

Darren159 committed:
Revert "Add support for saving and loading tasks from hard disk"

This reverts commit dad04a4ed73602684e0336391635e85c4a43a7f4.

Darren159 committed:
Revert "Revert "Add support for saving and loading tasks from hard disk""

This reverts commit 22a1825d49a869dfeb825a7ded17c17df7877bb1.

Darren159 committed:
Add support for Java LocalDateTime

Darren159 committed:
Merge branch 'branch-Level-8'

Darren159 committed:
Make the code more OOP

Darren159 committed:
Merge branch 'branch-A-MoreOOP'

Darren159 committed:
Divide classes into packages

Darren159 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Darren159 committed:
Add Gradle

Darren159 committed:
Merge branch 'branch-A-Gradle'

Darren159 committed:
Add JUnit tests

Darren159 committed:
Merge branch 'bracnh-A-JUnit'

Darren159 committed:
Add JavaDoc

Darren159 committed:
Add find by keyword

Darren159 committed:
Delete tasks.txt

Darren159 committed:
Merge branch 'branch-A-JavaDoc'

Darren159 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

Darren159 committed:
Add find by keyword

Darren159 committed:
Add JavaFX GUI

Darren159 committed:
Merge branch 'branch-Level-10'

licongshen12 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    protected String getResponse(String input) {
+        try {
+            Command c = Parser.parse(input);
+            if (c != null) {
+                return c.execute(tasks, ui, storage);
+            } else {
+                return ui.showInvalidCommand();
+            }
+        } catch (DukeException e) {
+            return ui.showError(e.getMessage());
+        }
+    }
```
I like how you name your methods, which in consistent through all the classes you created and explains clearly what the method is supposed to do.

licongshen12 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,120 @@
+    /**
+     * Enum to identify types of commands.
+     */
+    public enum CommandType {
+        LIST, DELETE, MARK, UNMARK, ADD, UNKNOWN, BYE, FIND
+    }
```
Good use of enums for CommandType!

dinde2004 replied:
I like the use of enums here.

licongshen12 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * The status of the task.
+     */
+    protected boolean isDone;
```
I like that your naming of booleans adheres to java coding standards!

licongshen12 commented on src/main/java/duke/exception/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Represents an exception that is thrown when a todo task has an empty description.
+ */
+public class EmptyDescriptionException extends DukeException {
+
+    /**
+     * Constructs a new EmptyDescriptionException with a default error message.
+     */
+    public EmptyDescriptionException() {
+        super("The description of a todo cannot be empty.");
+    }
+}
```
Good job abstracting out the different types of exceptions one might encounter into different classes!

licongshen12 reviewed (commented):
Your code was easy to understand due to the appropriate naming of methods and fields, as well as the Java documentation. LGTM!

dinde2004 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @return The command entered by the user.
+     */
+    public String readCommand() {
+        Scanner myObj = new Scanner(System.in);
+        return myObj.nextLine();
+    }
```
I think readCommand should be done in Parser class.

dinde2004 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+package duke.command;
```
I like the way you divide Command into multiple classes.

dinde2004 commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+package duke.exception;
```
I like the way you divide Exception types into multiple classes.

dinde2004 reviewed (commented):
Overall, your PR is very good.

Darren159 committed:
Duke, Storage, and TaskList classes: add assertions to safeguard against
null values

In the Duke, Storage, and TaskList classes, certain operations lack
validation checks to ascertain the non-null status of objects after
retrieval and addition operations, presenting the potential risk of
encountering NullPointerExceptions in the future.

Introduce assertions at key points in the Duke, Storage, and TaskList
classes. Specifically:

- In the `Duke` class, add an assertion post retrieval of the Command
object within the `getResponse` method to confirm it is not null.
- In the `Storage` class, assert the task list is not null after its
creation in the `load` method to prevent potential issues during file
reading operations.
- In the `TaskList` class, ensure that when a task is added to the list
through the `add` method, it is indeed present in the task list,
confirming successful addition.

Let's add assertions in the above-mentioned critical points to maintain
the robustness of the application by catching null pointer exceptions at
the earliest instance during the development phase.

Inserting these assertions allows for early detection and handling of
potential null pointer issues, thus aiding in maintaining a stable and
reliable codebase during development.

Darren159 committed:
Merge pull request #1 from Darren159/branch-A-Assertions

Duke, Storage, and TaskList classes: add assertions

Darren159 committed:
Merge branch 'master' of https://github.com/Darren159/ip

Darren159 committed:
Ui class: Remove unused methods and localize horizontalLine variable

The `Ui` class contains methods — `readCommand` and `showLine`, along
with the `horizontalLine` variable — that are not being utilized in the
application. The `horizontalLine` variable is also globally accessible
within the class, which is not necessary.

Unused methods and an unnecessarily globally accessible variable clutter
the codebase and make it less maintainable.

Remove the `readCommand` and `showLine` methods from the Ui class.
Additionally, change the scope of the `horizontalLine` variable to be
local, limiting its access to only the methods where it is utilized,
promoting a cleaner, more encapsulated code structure.

Let's remove the unused methods and restrict the `horizontalLine`
variable to a local scope to foster a cleaner, more maintainable
codebase.

Doing this removes dead code and prevents the unnecessary global
exposure of a variable, abiding by the best practices of clean code
which dictate that code should be as encapsulated as possible,
and that dead code should be removed.

Darren159 committed:
Merge pull request #2 from Darren159/branch-A-CodeQuality

Ui class: Remove unused methods and localize horizontalLine variable

Darren159 committed:
Merge branch 'master' of https://github.com/Darren159/ip

Darren159 committed:
Set up CI

Darren159 committed:
Use Streams in Ui

Darren159 committed:
Reorder import statements for test files

Darren159 committed:
Merge pull request #3 from Darren159/branch-A-Streams

Use Streams in Ui

Darren159 committed:
Merge branch 'master' of https://github.com/Darren159/ip

Darren159 committed:
Add Sort Deadlines function

Darren159 committed:
Merge pull request #4 from Darren159/branch-C-Sort

Add Sort Deadlines function

Darren159 committed:
Merge branch 'master' of https://github.com/Darren159/ip

Darren159 committed:
Add more error handling

Darren159 committed:
Add more error handling

Darren159 committed:
Merge branch 'A-MoreErrorHandling'

Darren159 committed:
Add welcome message

Darren159 committed:
Add more JUnit Tests

Darren159 committed:
Merge branch 'branch-A-MoreTesting'

Darren159 committed:
Add Product Screenshot

Darren159 committed:
Add User Guide

Darren159 committed:
Merge branch 'branch-A-UserGuide'

Darren159 committed:
Delete data folder

Darren159 committed:
Modify build name

Darren159 committed:
Modify User Guide

Darren159 committed:
Fix User Guide
